### ISSUE TITLE
Multiple version support of the operator

### ISSUE THREAD

**Issue by mdraijer** [CrunchyData/postgres-operator#2455]
First of all: it is possible that this feature is already supported but I'm not sure.
I know it was not supported when upgrading from 4.0.1 to 4.1.0 and I could not be sure from reading https://access.crunchydata.com/documentation/postgres-operator/4.6.1/upgrade/automatedupgrade/

## Overview

Make it possible to (temporarily) run multiple versions of the CrunchyData Postgres Operator in the same Kubernetes cluster, even if the CRD's differ.

## Use Case

To enable users of the Kubernetes cluster to upgrade their database clusters from the old to the new Operator version at their own pace.

## Desired Behavior

When upgrading from PGO 4.0.1 to 4.1.0 the CRD's changed their definition, but still had the same name, so it was not possible to run both versions in one Kubernetes cluster.
The CRD's from 4.1.0 onwards have a version attribute in the CRD, so maybe CrunchyData already takes care of this. But it is not clear whether you will indeed include a new version of the schema, with a new version number, every time that a new Operator version changes something in the definition.

Also it is important to know how many old version CrunchyData is planning to support in the CRD's. That determines how quickly users need to upgrade their databases to the new Operator version.

## Environment

- Platform: `Kubernetes`, `OpenShift`
- Platform Version: Kubernetes 1.19, Openshift 4.6
- PGO Image Tag: 4.6.1

**Comment by mdraijer** [comment]
I see from the CRD specs that it does adhere to versioning as described at e.g. https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/
So my question is mainly to express your intention to also _use_ that versioning strategy of the CRD's. And also how many old versions will be supported, i.e. when does a version get deprecated?

**Comment by jkatz** [comment]
Versioning is discussed in depth in a few places, including:

- [https://github.com/CrunchyData/postgres-operator/blob/master/CONTRIBUTING.md#releases--versioning](https://github.com/CrunchyData/postgres-operator/blob/master/CONTRIBUTING.md#releases--versioning)

The way that the CRD is presently structured, the CRD itself is compatible from 4.1 up to 4.7. That upgrade model may not always hold (i.e. we may move to a different upgrade model) but that does hold today. I will note that there was a big architecture shift in how HA was handled between 4.1 and 4.2, and we stopped making bug fixes for 4.1 more than a year ago.

For running "multiple versions of the Operator" -- generally Operators themselves are supposed to run as singletons. However, based on the current CRD structure, it is possible to run multiple versions of PGO within the same Kubernetes cluster, though you would need to ensure they are watching different namespaces. It may not also be that two Operators side-by-side work 100% as expected -- this is not a use case we fully account for or support -- but there are multiple instances we've seen of multiple PGO operators working side-by-side.

In terms of upgrading, I suggest poring over the [release notes](https://access.crunchydata.com/documentation/postgres-operator/latest/releases/). A lot of cool and practical things were added over the past 18 months since 4.1 was released, many of which improve the overall experience + make it easier to manage Postgres clusters (IMHO :smile:).
